src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingQueryMessageWireProtocol.cs

```diff
@@ -134,6 +152,71 @@ public TCommandResult Execute(IConnection connection, CancellationToken cancella
             }
         }
 
+        // private methods
+        private BsonDocument CombineCommandWithPayloads(ConnectionDescription connectionDescription)
+        {
+            if (_commandPayloads == null || _commandPayloads.Count == 0)
+            {
+                return _command;
+            }
+
+            var extraElements = new List<BsonElement>();
+            foreach (var type1Section in _commandPayloads)
+            {
+                var name = type1Section.Identifier;
+                var value = CreatePayloadArray(type1Section, connectionDescription);
+                var element = new BsonElement(name, value);
+                extraElements.Add(element);
+            }
+
+            var payloadAppendingSerializer = new ElementAppendingSerializer<BsonDocument>(BsonDocumentSerializer.Instance, extraElements);
+            return new BsonDocumentWrapper(_command, payloadAppendingSerializer);
+        }
+
+        private BsonArray CreatePayloadArray(Type1CommandMessageSection payload, ConnectionDescription connectionDescription)
+        {
+            IBsonSerializer payloadSerializer;
+            if (payload.Documents.CanBeSplit)
+            {
+                payloadSerializer = CreateSizeLimitingPayloadSerializer(payload, connectionDescription);
+            }
+            else
+            {
+                payloadSerializer = CreateFixedCountPayloadSerializer(payload);
+            }
+
+            var documents = new BsonDocumentWrapper(payload.Documents, payloadSerializer);
+            return new BsonArray { documents };
+        }
+
+        private IBsonSerializer CreateFixedCountPayloadSerializer(Type1CommandMessageSection payload)
+        {
+            var documentType = payload.DocumentType;
+            var serializerType = typeof(FixedCountBatchableSourceSerializer<>).MakeGenericType(documentType);
+            var itemSerializerType = typeof(IBsonSerializer<>).MakeGenericType(documentType);
+            var constructorParameterTypes = new[] { itemSerializerType, typeof(IElementNameValidator), typeof(int) };
+            var constructorInfo = serializerType.GetTypeInfo().GetConstructor(constructorParameterTypes);
+            var itemSerializer = payload.DocumentSerializer;
+            var itemElementNameValidator = payload.ElementNameValidator;
+            var count = payload.Documents.Count;
+            return (IBsonSerializer)constructorInfo.Invoke(new object[] { itemSerializer, itemElementNameValidator, count });
```

Comment:
I have a minor concern about using reflection here.

One alternative could be to add a CreatePayloadSerializer method to Type1CommandMessageSection.

But I'd rather keep Type1CommandMessageSection more of a data class than one that has behavior.

Thoughts?